File: src/utils/image_processing.rs
use crate::error::{BotError, BotResult};
use base64::{engine::general_purpose, Engine as _};
use bytes::Bytes;
use image::{imageops::FilterType, DynamicImage, ImageFormat};
use reqwest::Client;
use std::io::Cursor;
use tracing::info;

const MAX_SIZE_BYTES: usize = 10 * 1024 * 1024; // 10 MB
const MAX_DIMENSION: u32 = 1024; // Resize to max 1024x1024 to save space

/// Download an image, compress/resize it if it's over 10MB, and return it as a Base64 Data URI
pub async fn download_and_process_image(url: &str) -> BotResult<String> {
    let client = Client::new();
    let resp = client.get(url).send().await.map_err(|e| BotError::Custom(format!("Failed to download image: {}", e)))?;
    
    let content_type = resp.headers().get("content-type")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("image/png")
        .to_string();

    let bytes = resp.bytes().await.map_err(|e| BotError::Custom(format!("Failed to read image bytes: {}", e)))?;
    
    // If the image is already under 10MB, we can just return it directly as base64!
    if bytes.len() <= MAX_SIZE_BYTES {
        info!("Image is {} bytes (under 10MB), using directly.", bytes.len());
        let b64 = general_purpose::STANDARD.encode(&bytes);
        return Ok(format!("data:{};base64,{}", content_type, b64));
    }

    info!("Image is {} bytes (over 10MB). Resizing...", bytes.len());

    // It's over 10MB, we need to resize it.
    // If it's a GIF, resizing is complex, but we'll try to just load it as a static image for now
    // (A 10MB+ gif is huge, converting to static might be the only safe way).
    let img = image::load_from_memory(&bytes)
        .map_err(|e| BotError::Custom(format!("Failed to decode image for resizing: {}", e)))?;

    // Resize
    let resized = img.resize(MAX_DIMENSION, MAX_DIMENSION, FilterType::Lanczos3);

    // Save back to memory as JPEG (to save space) or PNG
    let mut out_bytes: Vec<u8> = Vec::new();
    let mut cursor = Cursor::new(&mut out_bytes);
    
    // Convert to JPEG to compress it heavily
    resized.write_to(&mut cursor, ImageFormat::Jpeg)
        .map_err(|e| BotError::Custom(format!("Failed to encode resized image: {}", e)))?;

    info!("Resized image to {} bytes.", out_bytes.len());

    if out_bytes.len() > MAX_SIZE_BYTES {
        return Err(BotError::Custom("Image is still too large even after resizing!".to_string()));
    }

    let b64 = general_purpose::STANDARD.encode(&out_bytes);
    Ok(format!("data:image/jpeg;base64,{}", b64))
}

File: ts/src/utils/imageProcessing.ts
/**
 * Image download + normalisation for the server-profile commands.
 *
 * Port of the Rust `src/utils/image_processing.rs`. Discord's `@me` guild-member
 * PATCH takes a base64 data URI, and rejects anything over 10 MB, so oversized
 * images are scaled to fit 1024×1024 and re-encoded as JPEG.
 *
 * Uses jimp (pure JavaScript) rather than a native codec — the Rust build could
 * not link `ring`/`cc` on this machine, and this port should not reintroduce a
 * toolchain dependency for something this small.
 */
import { Jimp } from 'jimp';
import { BotError, describe } from './../error';
import { logger } from './../logging';

const log = logger('image');

const MAX_SIZE_BYTES = 10 * 1024 * 1024; // 10 MB
const MAX_DIMENSION = 1024;
/** Matches the default quality of the Rust `image` crate's JPEG encoder. */
const JPEG_QUALITY = 75;

/**
 * Download an image, shrink it when it is over 10 MB, and return it as a base64
 * data URI ready for `PATCH /guilds/{id}/members/@me`.
 */
export async function downloadAndProcessImage(url: string): Promise<string> {
  let res: Response;
  try {
    res = await fetch(url);
  } catch (err) {
    throw BotError.custom(`Failed to download image: ${describe(err)}`);
  }

  // Rust skipped this check and would happily base64 an HTML error page.
  if (!res.ok) {
    throw BotError.custom(`Failed to download image: HTTP ${res.status}`);
  }

  const contentType = res.headers.get('content-type')?.split(';')[0]?.trim() || 'image/png';

  let bytes: Buffer;
  try {
    bytes = Buffer.from(await res.arrayBuffer());
  } catch (err) {
    throw BotError.custom(`Failed to read image bytes: ${describe(err)}`);
  }

  if (bytes.length <= MAX_SIZE_BYTES) {
    log.info(`Image is ${bytes.length} bytes (under 10MB), using directly.`);
    return `data:${contentType};base64,${bytes.toString('base64')}`;
  }

  log.info(`Image is ${bytes.length} bytes (over 10MB). Resizing...`);

  // An animated GIF loses its animation here — a 10 MB+ GIF cannot fit anyway.
  let out: Buffer;
  try {
    const image = await Jimp.fromBuffer(bytes);
    image.scaleToFit({ w: MAX_DIMENSION, h: MAX_DIMENSION });
    out = await image.getBuffer('image/jpeg', { quality: JPEG_QUALITY });
  } catch (err) {
    throw BotError.custom(`Failed to decode/encode image for resizing: ${describe(err)}`);
  }

  log.info(`Resized image to ${out.length} bytes.`);

  if (out.length > MAX_SIZE_BYTES) {
    throw BotError.custom('Image is still too large even after resizing!');
  }

  return `data:image/jpeg;base64,${out.toString('base64')}`;
}
